modin/core/dataframe/base/exchange/dataframe_protocol/dataframe.py

```diff
@@ -0,0 +1,543 @@
+         the purposes of this interchange protocol, or must producers either
+         reuse the chunk concept for this or copy the data?
+
+    Notes
+    -----
+    This ProtocolColumn object can only be produced by ``__dataframe__``,
+    so doesn't need its own version or ``__column__`` protocol.
+    """
+
+    @property
+    @abstractmethod
+    def size(self) -> int:
+        """
+        Size of the column, in elements.
+
+        Corresponds to `DataFrame.num_rows()` if column is a single chunk;
+        equal to size of this current chunk otherwise.
+
+        Returns
+        -------
+        int
+            Size of the column, in elements.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def offset(self) -> int:
+        """
+        Get the offset of first element.
+
+        May be > 0 if using chunks; for example for a column
+        with N chunks of equal size M (only the last chunk may be shorter),
+        ``offset = n * M``, ``n = 0 .. N-1``.
+
+        Returns
+        -------
+        int
+            The offset of first element.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def dtype(self) -> Tuple[DTypeKind, int, str, str]:
+        """
+        Dtype description as a tuple ``(kind, bit-width, format string, endianness)``.
+
+        * Kind : DTypeKind
+        * Bit-width : the number of bits as an integer
+        * Format string : data type description format string in Apache Arrow C
+                        Data Interface format.
+        * Endianness : current only native endianness (``=``) is supported
+
+        Returns
+        -------
+        tuple
+            ``(kind, bit-width, format string, endianness)``.
+
+        Notes
+        -----
+        - Kind specifiers are aligned with DLPack where possible
+          (hence the jump to 20, leave enough room for future extension).
+        - Masks must be specified as boolean with either bit width 1 (for bit masks)
+          or 8 (for byte masks).
+        - Dtype width in bits was preferred over bytes
+        - Endianness isn't too useful, but included now in case in the future
+          we need to support non-native endianness
+        - Went with Apache Arrow format strings over NumPy format strings
+          because they're more complete from a dataframe perspective
+        - Format strings are mostly useful for datetime specification, and for categoricals.
+        - For categoricals, the format string describes the type of the categorical
+          in the data buffer. In case of a separate encoding of the categorical
+          (e.g. an integer to string mapping), this can be derived from ``self.describe_categorical``.
+        - Data types not included: complex, Arrow-style null, binary, decimal,
+          and nested (list, struct, map, union) dtypes.
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def describe_categorical(self) -> Dict[str, Any]:
```

Comment:
do note that all existing protocols implement this _differently_, and I'm going to propose to change the spec - all protocols return a 3-element tuple here

Reply:
We will be matching the spec gradually. My thought is to introduce a base structure and correct details step-by-step.